api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go

```diff
@@ -0,0 +1,212 @@
+	config, err := PiccoloTargetProviderConfigFromMap(properties)
+	if err != nil {
+		return err
+	}
+	return d.Init(config)
+}
+func (s *PiccoloTargetProvider) SetContext(ctx *contexts.ManagerContext) {
+	s.Context = ctx
+}
+
+func (d *PiccoloTargetProvider) Init(config providers.IProviderConfig) error {
+	_, span := observability.StartSpan("Piccolo Target Provider", context.TODO(), &map[string]string{
+		"method": "Init",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	sLog.Info("  P (Piccolo Target): Init()")
+
+	// convert config to PiccoloTargetProviderConfig type
+	piccoloConfig, err := toPiccoloTargetProviderConfig(config)
+	if err != nil {
+		sLog.Errorf("  P (Piccolo Target): expected PiccoloTargetProviderConfig: %+v", err)
+		return err
+	}
+
+	d.Config = piccoloConfig
+	return nil
+}
+func toPiccoloTargetProviderConfig(config providers.IProviderConfig) (PiccoloTargetProviderConfig, error) {
+	ret := PiccoloTargetProviderConfig{}
+	data, err := json.Marshal(config)
+	if err != nil {
+		return ret, err
+	}
+	err = json.Unmarshal(data, &ret)
+	return ret, err
+}
+
+func (i *PiccoloTargetProvider) Get(ctx context.Context, deployment model.DeploymentSpec, references []model.ComponentStep) ([]model.ComponentSpec, error) {
+	ctx, span := observability.StartSpan("Piccolo Target Provider", ctx, &map[string]string{
+		"method": "Get",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	sLog.Infof("  P (Piccolo Target): getting artifacts: %s - %s, traceId: %s", deployment.Instance.Spec.Scope, deployment.Instance.ObjectMeta.Name, span.SpanContext().TraceID().String())
+
+	ret := make([]model.ComponentSpec, 0)
+	for _, component := range references {
+		name := component.Component.Name
+		req, err := http.NewRequest("GET", "http://0.0.0.0:9090/scenario/"+name, nil)
+		if err != nil {
+			continue
+		}
+		client := &http.Client{}
+		resp, err := client.Do(req)
+
+		if err != nil {
+			continue
+		}
+
+		defer resp.Body.Close()
+
+		switch resp.StatusCode {
+		case http.StatusOK:
+			// respBody, err := io.ReadAll(resp.Body)
+			component := model.ComponentSpec{
+				Name:       name,
+				Properties: make(map[string]interface{}),
+			}
+			component.Properties["workload.name"] = string(name)
+			ret = append(ret, component)
+		case http.StatusNotFound:
+			continue
+		}
+	}
+
+	return ret, nil
+}
+
+func (i *PiccoloTargetProvider) Apply(ctx context.Context, deployment model.DeploymentSpec, step model.DeploymentStep, isDryRun bool) (map[string]model.ComponentResultSpec, error) {
+	ctx, span := observability.StartSpan("Piccolo Target Provider", ctx, &map[string]string{
+		"method": "Apply",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	sLog.Infof("  P (Piccolo Target): applying artifacts: %s - %s, traceId: %s", deployment.Instance.Spec.Scope, deployment.Instance.ObjectMeta.Name, span.SpanContext().TraceID().String())
```

Comment:
Use sLog.InfofCtx(ctx, xxx) to always log the context.

Reply:
This commit was written too long ago, so the InfofCtx method does not exist.
I will create a new PR later to fix this.